File: its-my-training-day/leetcode-top-75/questions/graphs/dfs-bfs/clone-graph/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

class GraphNode:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else ArrayList()
def clone_graph(node: Optional[GraphNode]) -> Optional[GraphNode]:
    if node is None:
        return None
    clones: Dict[GraphNode, GraphNode] = {}

    def dfs(curr: GraphNode) -> GraphNode:
        if curr in clones:
            return clones[curr]
        copy = GraphNode(curr.val, ArrayList())
        clones[curr] = copy
        i = 0
        while i < curr.neighbors.size():
            copy.neighbors.add(dfs(curr.neighbors.get(i)))
            i += 1
        return copy

    return dfs(node)

File: its-my-training-day/leetcode-top-75/questions/graphs/dfs-bfs/clone-graph/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

export const cloneGraph = (node: GraphNode | null): GraphNode | null => {
  if (node === null) return null;
  const clones = new Map<GraphNode, GraphNode>();
  const dfs = (curr: GraphNode): GraphNode => {
    if (clones.has(curr)) return clones.get(curr) as GraphNode;
    const copy = new GraphNode(curr.val);
    clones.set(curr, copy);
    for (let i = 0; i < curr.neighbors.size(); i += 1) {
      copy.neighbors.add(dfs(curr.neighbors.get(i)));
    }
    return copy;
  };
  return dfs(node);
};
